velox/functions/sparksql/fuzzer/SparkQueryRunner.cpp

```diff
@@ -0,0 +1,388 @@
+#include "velox/functions/sparksql/fuzzer/SparkQueryRunner.h"
+#include "velox/functions/sparksql/fuzzer/proto/spark/connect/base.pb.h"
+#include "velox/functions/sparksql/fuzzer/proto/spark/connect/relations.pb.h"
+#include "velox/vector/arrow/Bridge.h"
+
+using namespace spark::connect;
+
+namespace facebook::velox::functions::sparksql::fuzzer {
+namespace {
+
+void writeToFile(
+    const std::string& path,
+    const std::vector<RowVectorPtr>& data,
+    memory::MemoryPool* pool) {
+  VELOX_CHECK_GT(data.size(), 0);
+
+  auto options = std::make_shared<dwio::common::WriterOptions>();
+  options->schema = data[0]->type();
+  options->memoryPool = pool;
+  // Spark does not recognize int64-timestamp written as nano precision in
+  // Parquet.
+  options->parquetWriteTimestampUnit = 6 /*kMirco*/;
+
+  auto writeFile = std::make_unique<LocalWriteFile>(path, true, false);
+  auto sink =
+      std::make_unique<dwio::common::WriteFileSink>(std::move(writeFile), path);
+  auto writer =
+      dwio::common::getWriterFactory(dwio::common::FileFormat::PARQUET)
+          ->createWriter(std::move(sink), options);
+
+  for (const auto& vector : data) {
+    // When vector is dictionary-encoded, complex types are not supported in
+    // exportFlattenedVector. Flatten the vector before writing to Parquet.
+    // https://github.com/facebookincubator/velox/issues/10397
+    VectorPtr flattened = vector;
+    BaseVector::flattenVector(flattened);
+    writer->write(flattened);
+  }
+  writer->close();
+}
+
+void appendComma(int32_t i, std::stringstream& sql) {
+  if (i > 0) {
+    sql << ", ";
+  }
+}
+
+std::string toCallSql(const core::CallTypedExprPtr& call);
+
+void toCallInputsSql(
+    const std::vector<core::TypedExprPtr>& inputs,
+    std::stringstream& sql) {
+  for (auto i = 0; i < inputs.size(); ++i) {
+    appendComma(i, sql);
+
+    const auto& input = inputs.at(i);
+    if (auto field =
+            std::dynamic_pointer_cast<const core::FieldAccessTypedExpr>(
+                input)) {
+      sql << field->name();
+    } else if (
+        auto call =
+            std::dynamic_pointer_cast<const core::CallTypedExpr>(input)) {
+      sql << toCallSql(call);
+    } else if (
+        auto lambda =
+            std::dynamic_pointer_cast<const core::LambdaTypedExpr>(input)) {
+      const auto& signature = lambda->signature();
+      const auto& body =
+          std::dynamic_pointer_cast<const core::CallTypedExpr>(lambda->body());
+      VELOX_CHECK_NOT_NULL(body);
+
+      sql << "(";
+      for (auto j = 0; j < signature->size(); ++j) {
+        appendComma(j, sql);
+        sql << signature->nameOf(j);
+      }
+
+      sql << ") -> " << toCallSql(body);
+    } else {
+      VELOX_NYI();
+    }
+  }
+}
+
+std::string toCallSql(const core::CallTypedExprPtr& call) {
+  std::stringstream sql;
+  sql << call->name() << "(";
+  toCallInputsSql(call->inputs(), sql);
+  sql << ")";
+  return sql.str();
+}
+
+std::string toAggregateCallSql(
```

Comment:
Wondering if this is a copy-paste from somewhere? If so, would it be possible to refactor to avoid code duplication?

Reply:
Yes, it is similar with Presto's toAggregateCallSql with one difference: Spark's aggregate does not support sort keys. I notice there are three `toAggregateCallSql` functions for Presto/DuckDB/Spark repectively, and wonder if we should refactor the sql generation for them to make some code reusable.

Reply:
@rui-mo It would be nice to consolidate this code.

Reply:
Extracting some reusable functions for SQL generation: https://github.com/facebookincubator/velox/pull/10545. Thanks.